index.html

```diff
@@ -1,390 +1,1451 @@
+            </ol>
+        </ol>
+
+      <p class="note">
+        The {{Promise}} returned by this algorithm is immediately resolved so
+        that it does not block canceling the stream. [[STREAMS]] specifies that
+        this algorithm will not be invoked again until a chunk is enqueued.
+      </p>
+
+      <li>
+        Let |cancelAlgorithm| be the following steps:
+        <ol>
+          <li>Let |promise| be [=a new promise=].
+          <li>
+            Run the following steps [=in parallel=].
+            <ol>
+              <li>Invoke the operating system to discard the contents of all
+              software and hardware receive buffers for the port.
+              <li>Invoke the steps to [=handle closing the readable stream=].
+              <li>
+                [=Queue a task=] to resolve |promise| on the [=serial port task
+                source=].
+            </ol>
+          <li>Return |promise|.
+        </ol>
+
+      <li>
+        [=ReadableStream/Set up=] |stream| with
+        <a href="https://streams.spec.whatwg.org/#readablestream-set-up-pullalgorithm">pullAlgorithm</a> set
+        to |pullAlgorithm| and
+        <a href="https://streams.spec.whatwg.org/#readablestream-set-up-cancelalgorithm">cancelAlgorithm</a> set
+        to |cancelAlgorithm|.
+      <li>Set [=this=].{{[[readable]]}} to |stream|.
+      <li>Return |stream|.
+    </ol>
+
+    To <dfn>handle closing the readable stream</dfn> perform the following
+    steps:
+
+    <ol>
+      <li>Set [=this=].{{[[readable]]}} to `null`.
+      <li>
+        If [=this=].{{[[writable]]}} is `null` and
+        [=this=].{{[[pendingClosePromise]]}} is not `null`, [=queue a task=] to
+        [=resolve=] [=this=].{{[[pendingClosePromise]]}} with `undefined` on the
+        [=serial port task source=].
+    </ol>
+
+  </section>
+
+  <section>
+    <h3><dfn>writable</dfn> attribute</h3>
+
+    <aside class="example">
+      To write individual chunks of data to the port a
+      {{WritableStreamDefaultWriter}} can be created and released as necessary.
+      This example uses a `TextEncoder` to encode a {{DOMString}} as the
+      necessary {{Uint8Array}} for transmission.
+
+      <pre class="js">
+        const encoder = new TextEncoder();
+        const writer = port.writable.getWriter();
+        await writer.write(encoder.encode("PING"));
+        writer.releaseLock();
+      </pre>
+
+      When writing larger chunks it can be important to allow the port to apply
+      back pressure so that the serial transmitter does not get too far behind
+      sending data generated by the application. The
+      {{WritableStreamDefaultWriter/write()}} method returns a {{Promise}} which
+      resolves when data has been written. While having some data available in
+      the transmit buffer is important to maintain good throughput awaiting this
+      {{Promise}} before generating too many chunks of data is a good practice
+      to avoid excessive buffering.
```

Comment:
This might also be a job for the `writer.ready` promise? That reflects the high water mark, which you are currently leaving at its default of 1, but maybe you could let the creator of the serial port customize it?

Reply:
As I mentioned in my last comment in Chromium the underlying sink is implemented using a buffer of `options["bufferSize"]` bytes into which chunks are written. Exposing the amount of free space in this buffer as `writer.desiredSize` could be useful. I set the high water mark to 1 because, given the presence of this buffer, I don't want to have the stream itself form an additional queue in which data can be buffered.

Reply:
Hmm I see. I agree we don't want that extra buffer, but it sure would be nice if that was exposed.

On the implementation side, I wonder if we can have a "fake HWM", that basically says "I'm managing the buffer elsewhere, outside the streams infrastructure, but I'm going to tell you how much space is in it." /cc @ricea

On the specification side, I think we could get away with just setting the HWM to an implementation-defined value (like I suggest for the readable side spec above). The fact that the queue is managed elsewhere is an implementation detail, perhaps worth mentioning in a note, but not observable to JS.

Reply:
> On the implementation side, I wonder if we can have a "fake HWM", that basically says "I'm managing the buffer elsewhere, outside the streams infrastructure, but I'm going to tell you how much space is in it." /cc @ricea

I've thought about a way to override `desiredSize` to sever the connection between it and the internal queue. Currently all the platform-created streams I know of use a HWM of 0 or 1 to avoid buffer bloat. So `desiredSize` isn't fulfilling its original purpose.

I don't have anything like a detailed design.